enhancements/kubelet/node-swap.md

```diff
@@ -0,0 +1,142 @@
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The upstream Kubernetes 1.22 release introduced alpha support for configuring swap memory usage for Kubernetes workloads on a per-node basis.
+
+Now that swap use on nodes is supported in upstream, there are a number of use cases that would benefit from OpenShift nodes supporting swap, including improved node stability, better support for applications with high memory overhead but smaller working sets, the use of memory-constrained devices, and memory flexibility.
+
+## Motivation
+
+See [KEP-2400: Motivation].
+
+[KEP-2400: Motivation]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#motivation
+
+### Goals
+
+- Swap can be provisioned and configured for nodes to use in an OpenShift cluster.
+
+### Non-Goals
+
+- Workload-specific swap accounting.
+- Any of the non-goals in [KEP-2400: Non-goals].
+
+[KEP-2400: Non-goals]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#non-goals
+
+## Proposal
+
+### User Stories
+
+See [KEP-2400: User Stories](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#user-stories).
+
+### Implementation Details/Notes/Constraints [optional]
+
+See [KEP-2400: Notes/Constraints/Caveats](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#notesconstraintscaveats-optional).
+
+### Risks and Mitigations
+
+See [KEP-2400: Risks and Mitigations].
+
+[KEP-2400: Risks and Mitigations]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#risks-and-mitigations
+
+## Design Details
+
+_very drafty_
+
+- We will add `NodeSwap` to the [`TechPreviewNoUpgrade`] feature gate list.
+- We can use [ignition configs](https://coreos.github.io/ignition/configuration-v3_3/) to add swap partitions to worker nodes. (filesystems.format = swap)
+  - Options: partition on the root node, or perhaps provision and mount an NVMe volume? (https://github.com/openshift/machine-config-operator/issues/1619)
+- We will need to ensure the node has `swapon` before kubelet starts
+- The kubelet needs an appropriate KubeletConfiguration (e.g. `NodeSwap` feature flag enabled, `failSwapOn = false`, and [`memorySwap.SwapBehavior` set](https://kubernetes.io/docs/concepts/architecture/nodes/#swap-memory))
+
+TODO: how will we modify MCO to roll this out? Do we need to at all?
+
+### Open Questions [optional]
+
+- Will we eventually want to enable swap on all OpenShift nodes by default?
+- Should swap just be limited to worker nodes, or should we consider adding it to control plane nodes too?
```

Comment:
Maybe just worker nodes, but allow swap to be enabled on masters? We probably need some testing around this.